contrib/epee/include/storages/portable_storage_from_json.h

```diff
@@ -128,20 +128,20 @@ namespace epee
                   errno = 0;
                   int64_t nval = strtoll(val.data(), NULL, 10);
                   if (errno) throw std::runtime_error("Invalid number: " + std::string(val));
-                  stg.set_value(name, nval, current_section);              
+                  stg.set_value(name, int64_t(nval), current_section);
```

Comment:
Why is that needed since nval is already int64_t ?

Reply:
The `static_assert` within `set_value` will complain about copying a value. Alternatively, I could change the check to ignore all `is_arithmetic` copies since the intent is to catch more complex copy functions.

Reply:
I don't mind, I just wanted to understand. Does the cast make this expression a temporary, technically ?

Reply:
Yes.

Reply:
It's an unnecessary change that makes parts of the code like this one worse.  It should be left up to the caller to decide whether passing an rvalue reference or making a copy is appropriate, and the `static_assert` should be a tool used to track down expensive copies *in a local branch* (or with some off-by-default local define guard), not something forced on all users of the interface.  The caller always has more information than the callee, and forcing the caller to make explicit copies everywhere they want to do so because someone somewhere *might* use a copy when they shouldn't is bad interface design.

Reply:
It actually makes things worse. The change utilizes intermediate CPU register for storing a value and then saving it to the stack instead of just using the stack.

Tested on MinGW 9.2.0 with default release compilation flags (i.e. -O2). 

Before:
```asm
	call	strtoll
	movq	%rax, 112(%rsp)
	call	*%r12
	movl	(%rax), %ecx
	testl	%ecx, %ecx
	jne	.L46392
	movq	72(%rsp), %rdx
	movq	%rbp, %r9
	leaq	112(%rsp), %r8
	movq	%rdi, %rcx
	call	_ZN4epee13serialization16portable_storage9set_valueIRxEEbRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEOT_PNS0_7sectionE
```

After:
```asm
	call	strtoll
	movq	%rax, %r13
	call	*%r12
	movl	(%rax), %ecx
	testl	%ecx, %ecx
	jne	.L46293
	movq	72(%rsp), %rdx
	movq	%rbp, %r9
	leaq	112(%rsp), %r8
	movq	%rdi, %rcx
	movq	%r13, 112(%rsp)
	call	_ZN4epee13serialization16portable_storage9set_valueIxEEbRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEOT_PNS0_7sectionE
```

Reply:
> It's an unnecessary change that makes parts of the code like this one worse. It should be left up to the caller to decide whether passing an rvalue reference or making a copy is appropriate, and the static_assert should be a tool used to track down expensive copies in a local branch (or with some off-by-default local define guard), not something forced on all users of the interface. The caller always has more information than the callee, and forcing the caller to make explicit copies everywhere they want to do so because someone somewhere might use a copy when they shouldn't is bad interface design.

The callee (in `portable_storage`) **always** needs a copy of the value in this function. Its not obvious due to the way it is currently written (there is some outstanding inefficient map usage to be handled later). The function could take the argument by value and then move (as per recommendation by Scott Meyers), however in this situation my preference was to force copying to be explicit by the caller similar to Rust. Implicit copies can be kind of frustrating in this type of code (IMO at least).

----------------- 

@xiphon I don't see much of interest from the diassembly, except for odd output differences from mingw. The first version takes the output from `stroll` and immediately shoves it onto the stack for use in the `set_value` function which needs a reference to the value. In the second case, it shoves the output from `stroll` into `r13` and then takes the value from `r13` and shoves it onto the stack like the first version (which still needs a reference unfortunately due to C++ r-value rules).

There is a lesser-known library called `boost::call_traits` or something. It automatically drops references to arithmetic template arguments so when not-inlined, the compiler can pass the argument via register, etc. Not sure if it was updated for r-value stuff. But it would also mean not having to manually copy arithmetic values.

Reply:
The interesting part is what it does with std::string.

Reply:
@xiphon I think you may have been pointing out that the extra "copy" here of the `int64_t` caused the output to change ... ? The output from gcc 9.2 target the Linux x86-64 doesn't do this extra work, the `rax` register is shoved onto the stack exactly once and everything is setup as expected. So I'm not sure what to do say about the mingw output ...

Reply:
> The function could take the argument by value and then move (as per recommendation by Scott Meyers), however in this situation my preference was to force copying to be explicit by the caller similar to Rust.

Can you elaborate on why you think it is better to import a foreign paradigm from another programming language than to listen to the advice of one of the most experienced, influential people in the C++ community?

Reply:
@vtnerd 
Same happens on Linux.

@moneromooo-monero 
> The interesting part is what it does with std::string.

It does an extra copy with `t_type(d)`.

Debian GCC 9.2, default release compilation flags (i.e. `-Ofast`).

Without `t_type(d)`, i.e. `return stg.set_value(pname, d, hparent_section);`:
```asm
_ZN4epee13serialization58kv_serialization_overloads_impl_is_base_serializable_typesILb1EE12kv_serializeINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEENS0_16portable_storageEEEbRKT_RT0_NSE_8hsectionEPKc:
.LFB57876:
	.cfi_startproc
	.cfi_personality 0x9b,DW.ref.__gxx_personality_v0
	.cfi_lsda 0x1b,.LLSDA57876
	endbr64
	pushq	%r15
	.cfi_def_cfa_offset 16
	.cfi_offset 15, -16
	pushq	%r14
	.cfi_def_cfa_offset 24
	.cfi_offset 14, -24
	movq	%rdx, %r14
	movq	$-1, %rdx
	pushq	%r13
	.cfi_def_cfa_offset 32
	.cfi_offset 13, -32
	movq	%rdi, %r13
	pushq	%r12
	.cfi_def_cfa_offset 40
	.cfi_offset 12, -40
	movq	%rsi, %r12
	pushq	%rbp
	.cfi_def_cfa_offset 48
	.cfi_offset 6, -48
	movq	%rcx, %rbp
	pushq	%rbx
	.cfi_def_cfa_offset 56
	.cfi_offset 3, -56
	subq	$56, %rsp
	.cfi_def_cfa_offset 112
	movq	%fs:40, %rax
	movq	%rax, 40(%rsp)
	xorl	%eax, %eax
	leaq	16(%rsp), %rbx
	movq	%rbx, (%rsp)
	testq	%rcx, %rcx
	je	.L42377
	movq	%rcx, %rdi
	call	strlen@PLT
	leaq	0(%rbp,%rax), %rdx
.L42377:
	movq	%rsp, %r15
	movq	%rbp, %rsi
	movq	%r15, %rdi
.LEHB3709:
	call	_ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE12_M_constructIPKcEEvT_S8_St20forward_iterator_tag.constprop.0
.LEHE3709:
	movq	%r14, %rcx
	movq	%r13, %rdx
	movq	%r15, %rsi
	movq	%r12, %rdi
.LEHB3710:
	call	_ZN4epee13serialization16portable_storage9set_valueIRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEbSA_OT_PNS0_7sectionE@PLT
```

With `t_type(d)`, i.e. `return stg.set_value(pname, t_type(d), hparent_section);`:
```asm
_ZN4epee13serialization58kv_serialization_overloads_impl_is_base_serializable_typesILb1EE12kv_serializeINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEENS0_16portable_storageEEEbRKT_RT0_NSE_8hsectionEPKc:
.LFB57876:
	.cfi_startproc
	.cfi_personality 0x9b,DW.ref.__gxx_personality_v0
	.cfi_lsda 0x1b,.LLSDA57876
	endbr64
	pushq	%r15
	.cfi_def_cfa_offset 16
	.cfi_offset 15, -16
	pushq	%r14
	.cfi_def_cfa_offset 24
	.cfi_offset 14, -24
	pushq	%r13
	.cfi_def_cfa_offset 32
	.cfi_offset 13, -32
	pushq	%r12
	.cfi_def_cfa_offset 40
	.cfi_offset 12, -40
	movq	%rdx, %r12
	pushq	%rbp
	.cfi_def_cfa_offset 48
	.cfi_offset 6, -48
	movq	%rcx, %rbp
	pushq	%rbx
	.cfi_def_cfa_offset 56
	.cfi_offset 3, -56
	subq	$104, %rsp
	.cfi_def_cfa_offset 160
	movq	8(%rdi), %rdx
	movq	%rsi, 8(%rsp)
	movq	(%rdi), %rsi
	leaq	16(%rsp), %r15
	leaq	32(%rsp), %rbx
	movq	%r15, %rdi
	leaq	64(%rsp), %r13
	addq	%rsi, %rdx
	movq	%fs:40, %rax
	movq	%rax, 88(%rsp)
	xorl	%eax, %eax
	movq	%rbx, 16(%rsp)
.LEHB3615:
	call	_ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE12_M_constructIPcEEvT_S7_St20forward_iterator_tag.constprop.0
.LEHE3615:
	movq	%r13, 48(%rsp)
	movq	$-1, %rdx
	testq	%rbp, %rbp
	je	.L41695
	movq	%rbp, %rdi
	call	strlen@PLT
	leaq	0(%rbp,%rax), %rdx
.L41695:
	leaq	48(%rsp), %r14
	movq	%rbp, %rsi
	movq	%r14, %rdi
.LEHB3616:
	call	_ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE12_M_constructIPKcEEvT_S8_St20forward_iterator_tag.constprop.0
.LEHE3616:
	movq	8(%rsp), %rdi
	movq	%r12, %rcx
	movq	%r15, %rdx
	movq	%r14, %rsi
.LEHB3617:
	call	_ZN4epee13serialization16portable_storage9set_valueINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEbRKS8_OT_PNS0_7sectionE@PLT
```

Reply:
> Can you elaborate on why you think it is better to import a foreign paradigm from another programming language than to listen to the advice of one of the most experienced, influential people in the C++ community?

Yes, as already stated, I am trying to prevent implicit copies from occurring. It is extremely easy to miss at the call site, especially since the compiler will not "trap" or warn the user if the function parameters change from `const&` to by-value. So these checks partially helped me track down where copies occurred. So I think the checking could continue helping with maintenance. These are very frequently called functions within the codebase. And lastly, explicitly moving or requesting a copy at the call site is a trivial amount of typing and affects a tiny section of the codebase since all of this is hidden when adding fields to a p2p or rpc call.

-------------------

>  @vtnerd
> Same happens on Linux.

You've changed to an entirely differently function here -

This forcibly put a `std::string` copy into this function from another function. I still don't see your point - the copy occurs either way because `portable_storage` design requires ownership of the `std::string` from the struct that is going to be serialized. In other words, my changes add an explicit copy inside of this function instead of an implicit copy hidden within `set_value`. Any complaints about copying a `std::string` in this situation are primarily due to the existing design that I am hoping to slowly improve.

Reply:
Yes, i have to correct myself, it doesn't make things worse from memory management perspective, i.e. no extra copy should occur because 
> changes add an explicit copy inside of this function instead of an implicit copy hidden within `set_value`. Any complaints about copying a `std::string` in this situation are primarily due to the existing design that I am hoping to slowly improve.

Though every time i look at this code line it makes me wonder what is going on here and why do we now have an argument copy here.